persistence-api/src/main/java/io/stargate/db/limiter/AsyncRateLimiter.java

```diff
@@ -0,0 +1,218 @@
+   *     capacity.
+   * @param reserveWindowUnit time unit for the reserve window
+   */
+  public AsyncRateLimiter(
+      ScheduledExecutorService executor,
+      long rate,
+      TimeUnit rateUnit,
+      long reserveWindow,
+      TimeUnit reserveWindowUnit) {
+    this(executor, rate, rateUnit, reserveWindow, reserveWindowUnit, 1, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Update the rate of the this limiter.
+   *
+   * @param rate the new rate (number of permits to allow per {@code rateUnit}).
+   * @param rateUnit time unit for {@code rate}.
+   */
+  public void setRate(long rate, TimeUnit rateUnit) {
+    this.nanosPerPermit = rateUnit.toNanos(1) * 1.0 / rate;
+  }
+
+  /**
+   * The rate of this limiter.
+   *
+   * @param rateUnit the time unit in which to return the rate.
+   * @return this limiter rate (number of permits allowed per {@code rateUnit}).
+   */
+  public long getRate(TimeUnit rateUnit) {
+    return (long) (rateUnit.toNanos(1) * 1.0 / nanosPerPermit);
+  }
+
+  /**
+   * Acquire the given number of permits and return the time in nanoseconds for which the works
+   * should be scheduled.
+   */
+  private long acquire(long permits, long currentTimeNanos) {
+    // Do not delay if no permits are requested, even if late. 0 work is already accounted for in
+    // the previous
+    // acquire call.
+    if (permits <= 0) return currentTimeNanos;
+
+    long scheduleTime;
+    long timeToAcquire = (long) (permits * nanosPerPermit);
+
+    while (true) {
+      long consumedTo = consumedToTime.get();
+      // scheduleTime is the time at which the task is allowed to start, which can be both in the
+      // future (if running at limit) or in the past.
+      // Make sure it's not more than the reserve window in the past though.
+      scheduleTime = Math.max(consumedTo, currentTimeNanos - reserveWindowNanos);
+      if (consumedToTime.compareAndSet(consumedTo, scheduleTime + timeToAcquire)) {
+        return scheduleTime;
+      } // Else we have had a concurrent modification. Retry.
+    }
+  }
+
+  /**
+   * Reserves the given number of permits and executes the provided asynchronous task when they
+   * become available.
+   *
+   * <p>This may mean executing the task immediately on the current thread (if permits are already
+   * available) or picking a time for which it is scheduled.
+   *
+   * @param permits the number of permits to acquire.
+   * @param task an asynchronous task.
+   * @return returns the number of nanoseconds the task has to wait before being executed.
+   */
+  public <T> CompletableFuture<T> acquireAndExecute(
+      long permits, Supplier<CompletableFuture<T>> task) {
+    long currentTime = System.nanoTime();
+    long scheduleTime = acquire(permits, currentTime);
+    long delay = scheduleTime - currentTime;
+
+    if (delay < schedulingThresholdNanos) {
+      // Time is in the past, or very close in the future. Execute immediately.
+      return task.get();
```

Comment:
nit: To me personally `.get()` looks like a method that should generally be idempotent. I know that the `Supplier` contract does not impose any requirements on the implementation, still it feels a bit odd to me. What do you think about using `Callable` for the `task` parameter?

Reply:
Honestly, `Callable` is more annoying because of it throwing `Exception`. Here, I can't propagate the exception up (because the `Persistence` methods don't throw `Exception` in particular) so we'd end up wrapping it in a `RuntimeException` or equivalent. To me, `Supplier` exists exactly to avoid that annoyance of `Callable` coming from a per-lambda, not to add some implicit idempotency contract, and I've see tons of use of it not idempotent so I'd be careful with running with that assumption.

So I haven't changed it, but that said, not trying to be difficult: if it bugs you enough that you'd prefer using `Callable` despite the need to sometimes wrap the resulting exception, then let me know and I'll change.

Reply:
No worries. I too thought exceptions would be annoying, but have not tried to actually refactor :) Current code is good from my POV.